server/build-plugins.js

```diff
@@ -38,12 +38,22 @@ const prepareModules = (plugins = [], local = true) => {
   if (installPackages.length) {
     try {
       logger.info('Installing plugin packages...');
-      execSync(
-        `npm install --no-save ${installPackages.join(' ')} --production`,
-        {
-          stdio: [0, 1, 2]
-        }
-      );
+      if (!local) {
+        // check if connected to internet
+        // if not, skip npm install
+        await require('dns').lookup('https://purse.io', err => {
```

Comment:
Heh, was wondering if I'd get called out on the endpoint. I just didn't like the idea of using someone else's endpoint for our check 🤷‍♂️. I'm not sure if pointing to the DNS is even that reliable. The suggestions I was seeing online had the url rather than an IP or DNS, I'm guessing because usually you have backups (`8.8.4.4` for google) when setting DNS' as far as I know. Also, pointing to google's dns' would cause problems for users in places like China. I'm not sure how npm is treated there though so it might not make a difference. Either way, makes me think that checking if npmjs.com is available since that's all we need to hit in the next line. 

And yeah, I'd rather not have to spin up a child process since it's not really necessary. Is there a specific reason to use `ping` rather than `lookup`?

Reply:
Good call on Google's DNS being blocked in China.

The reason why I say using `ping` is better than using a DNS lookup is the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege) - we don't really need the information from a DNS lookup, we just need to know if we can reach out to the internet.

I think that the URI that we reach out to needs to be configurable via an env variable in the case 
 that the site is down or the user wants more control or the site is blocked in a country, we could provide a default but i think it should be something that is highly available across the world, a user can still automate the process by passing an env variable in edge cases

Reply:
I think the lookup for npm should solve for this. We're not really testing for internet access but rather whether they can download from npm.

Reply:
`least privilege` seems to be more about what the server should allow to be requested than what the client should be requesting. Or at least that's how I'm reading it.

Reply:
What if the user is using plugins from a private npm repo?

Reply:
I think that's irrelevant to this section since this is just for plugin install from the `pluginsConfig` and run pre-webpack build. It takes the name from the config and adds it to the npm install. We only support direct npm repo and github repo right now (which downloads from github if in this format `my-org/my-repo`). For private repos, my understanding is that that is set in the package.json which wouldn't be affected by this anyway.

Reply:
Like this:

```
$ GIT_CURL_VERBOSE=1 git pull
```

You don't need to set the environment variable, it works without it. If the value is hardcoded in, the user will need to edit code when there are edge cases around the DNS lookup not working when it should, which does not lend itself towards automation. Something as simple as:

```
const EXTERNAL_URI = process.env.EXTERNAL_URI || 'https://www.npmjs.com';
```

Reply:
If you have a strong preference to hardcode in the value then I suppose that is fine but I do not think that it is a good idea because it will make automation more difficult. If I wanted to write a script to set up bPanel and didn't want to rely on the fact that npm has a DNS entry then I could have the option to choose who I rely on

Reply:
Ah ok I see what you mean! Not that I feel strongly about hard coding it, I just didn't want to overcomplicate configuration for a small edge case. This is fine.